File: C/Algorithms/Searching/Binary_Search.c
#include<stdlib.h>
#include<stdio.h>

int BinarySearch(int arr[], int len, int elem);

int main()
{
    int i,elem,len=0;

    printf("\n\n\n\t\t\t\tBINARY SEARCH");
    printf("\n\nEnter array length : ");
    scanf("%d",&len);                       //inputting aray size

    int *arr = (int*) calloc(len,sizeof(int));

    printf("\n\nEnter array elements : ");
    for(i=0;i<len;i++){
       scanf("%d",&arr[i]);                 //inputting aray elements and array must be sorted to perform binary search
    }

    printf("\n\nEnter element to be searched : ");
    scanf("%d",&elem);                      //inputting element to be searched

    int foundAt = BinarySearch(arr,len,elem);

    if(foundAt==-1)
        printf("\n\n\t\tGiven element not found in the given array!");
    else
        printf("\n\n\t\tGiven element found in the given array at %d position",foundAt+1);

    return 0;
}

int BinarySearch(int arr[], int len, int elem){
    int beg=0;
    int last=len-1;
    int mid;
    while(beg<=last){
        mid=(beg+last)/2;
        if(arr[mid] == elem)
            return mid;
        else if(arr[mid] > elem)
            last=mid-1;
        else
            beg=mid+1;
    }
    return -1;
} 

File: Rust/Algorithms/Searching/binary_search.rs
fn main() {
    let v = vec![2, 5, 6, 7, 8, 9, 12, 45, 65, 66, 85];
    let index = binary_search(&v, 100); 
    if index != -1 {
        println!("Element is found at index {}", index);
    } else {
        println!("Element not found in the vector");
    }
}

// Recursive Binary Search
fn binary_search_rec(vector: &Vec<i32>, l: usize, r: usize, val: i32) -> i32 {
    if l >= r {
        return -1
    }

    let mid = (l + r) / 2;
    if vector[mid] == val {
        return mid as i32
    }

    if vector[mid] > val {
        return binary_search_rec(vector, l, mid-1, val)
    } else {
        return binary_search_rec(vector, mid+1, r, val)
    }
}

// Iterative Binary Search
fn binary_search(vector: &Vec<i32>, val: i32) -> i32 {
    let mut l = 0;
    let mut r = vector.len() - 1;

    while l <= r {
        let mid = (l + r) / 2;
        if vector[mid] == val {
            return mid as i32
        }

        if vector[mid] > val {
            r = mid - 1;
        } else {
            l = mid + 1;
        }
    }
    return -1
}